src/main/java/seedu/address/logic/commands/ChangeCommand.java

```diff
@@ -0,0 +1,204 @@
+            + PREFIX_TIME + "4:00PM"
+            + PREFIX_DURATION + "30";
+
+    public static final String MESSAGE_EDIT_APPOINTMENT_SUCCESS = "Edited Appointment: %1$s";
+    public static final String MESSAGE_DUPLICATE_APPOINTMENT = "This appointment already exists in the appointment "
+            + "book.";
+
+    public static final String ASSIGNMENT_OVERLAP = "This time slot is occupied";
+    public static final String DATE_MISSING = "The date of appointment is missing";
+    public static final String TIME_MISSING = "The time of appointment is missing";
+    public static final String DURATION_MISSING = "The duration of appointment is missing";
+
+    private final Index index;
+    private final EditAppointmentDescriptor editAppointmentDescriptor;
+
+    /**
+     * @param index of the appointment in the filtered appointment list to edit
+     * @param editAppointmentDescriptor details to edit the appointment with
+     */
+    public ChangeCommand(Index index, EditAppointmentDescriptor editAppointmentDescriptor) {
+        requireNonNull(index);
+        requireNonNull(editAppointmentDescriptor);
+
+        this.index = index;
+        this.editAppointmentDescriptor = new EditAppointmentDescriptor(editAppointmentDescriptor);
+    }
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        List<Appointment> lastShownList = model.getFilteredAppointmentList();
+
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_APPOINTMENT_DISPLAYED_INDEX);
+        }
+
+        Appointment appointmentToEdit = lastShownList.get(index.getZeroBased());
+        Appointment editedAppointment = createEditedAppointment(appointmentToEdit, editAppointmentDescriptor);
+
+        if (appointmentToEdit.equals(editedAppointment) && model.hasAppointment(editedAppointment)) {
+            throw new CommandException(MESSAGE_DUPLICATE_APPOINTMENT);
+        }
+        model.deleteAppointment(appointmentToEdit);
+        if (model.hasOverlappingAppointment(editedAppointment)) {
+            model.addAppointment(appointmentToEdit);
+            throw new CommandException(ASSIGNMENT_OVERLAP);
+        }
+        model.addAppointment(appointmentToEdit);
+        model.setAppointment(appointmentToEdit, editedAppointment);
+        model.updateFilteredAppointmentList(PREDICATE_SHOW_ALL_APPOINTMENTS);
+        return new CommandResult(String.format(MESSAGE_EDIT_APPOINTMENT_SUCCESS, editedAppointment));
+    }
+
+    /**
+     * Creates and returns an {@code Appointment} with the details of {@code appointmentToEdit}
+     * edited with {@code editAppointmentDescriptor}.
+     */
+    private static Appointment createEditedAppointment(Appointment appointmentToEdit,
+                                                       EditAppointmentDescriptor editAppointmentDescriptor) {
+        assert appointmentToEdit != null;
```

Comment:
Beautiful assertion!

Reply:
HAHA thanks :)

Reply:
I have to say it too!
It solves the warning from `get()` of `Optional`!
Thanks for the enlightenment.